kernel/src/platform/mod.rs

```diff
@@ -139,6 +139,56 @@ pub trait Chip {
     unsafe fn print_state(&self, writer: &mut dyn Write);
 }
 
+/// Interface for handling interrupts and deferred calls on a hardware chip.
+///
+/// Each board must implement this trait to handle specific
+/// interrupts. When an interrupt (identified by number) has triggered and
+/// should be handled, the implementation of this trait will be called with the
+/// interrupt number. The implementation can then handle the interrupt, or
+/// return `false` to signify that it does not know how to handle the interrupt.
+///
+/// This functionality is given this `InterruptService` interface so that
+/// multiple objects can be chained together to handle interrupts for a chip.
+/// This is useful for code organization and removing the need for duplication
+/// when multiple variations of a specific microcontroller exist. Then a shared,
+/// base object can handle most interrupts, and variation-specific objects can
+/// handle the variation-specific interrupts.
+///
+/// To simplify structuring the Rust code when using `InterruptService`, the
+/// interrupt number should be passed "top-down". That is, an interrupt to be
+/// handled will first be passed to the `InterruptService` object that is most
+/// specific. If that object cannot handle the interrupt, then it should
+/// maintain a reference to the second most specific object, and return by
+/// calling to that object to handle the interrupt. This continues until the
+/// base object handles the interrupt or decides that the chip does not know how
+/// to handle the interrupt. For example, consider a `nRF52840` chip that
+/// depends on the `nRF52` crate which in turn depends on the `nRF5` crate. If
+/// all three have specific interrupts they know how to handle, the flow would
+/// look like:
+///
+/// ```ignore
+///           +---->nrf52840
+///           |        |
+///           |        |
+///           |        v
+///           |      nrf52
+///           |        |
+///           |        |
+///           |        v
+/// kernel-->nrf5     nrf5
+/// ```
+/// where the kernel instructs the `nrf5` crate to handle interrupts, and if
+/// there is an interrupt ready then that interrupt is passed through the crates
+/// until something can service it.
+pub trait InterruptService<T> {
```

Comment:
Shouldn't `T` be bounded by some kind of "task" trait?

Reply:
That would make sense, but we don't have any `DeferredCallTask` trait today.

Reply:
Shouldn't we introduce one with this pull-request then? 
- If we always use `T = DeferredCallTask` for now, I don't see the need for it to be generic.
- If we use various task types, I think we should unify them with a trait.

Reply:
Each chip defines its own `Task` type (sam4l calls it `Task`, nrf52 calls it `DeferredCallTask`), so it has to be a generic or an associated type for it to be part of the `InterruptService` trait definition. Perhaps an associated type would be better?

Currently, each `Task` type is just an enum that specifies different task types and can be converted to/from usize. Given that there are no common functions implemented across all `Task` types I am not sure how much value is gained from making a trait with no functions. 

Would adding the last line of the following code snippet really help that much with unifying?

```rust
/// A type of task to defer a call for
#[derive(Copy, Clone)]
pub enum DeferredCallTask {
    Nvmc = 0,
}

impl DeferredTaskTrait for DeferredCallTask {}
```

Reply:
I think such a marker trait would be useful, in restricting what kind of type can be applied there. Besides, when someone implements a new chip, they can look where `DeferredTaskTrait` is already implemented and more quickly understand what to do.